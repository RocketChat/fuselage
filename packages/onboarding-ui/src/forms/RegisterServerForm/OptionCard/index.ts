import List from '../../../common/List';
import Block from './Block';
import Description from './Description';
import OptionCard from './OptionCard';
import Subtitle from './Subtitle';
import Title from './Title';

export default Object.assign(OptionCard, {
  Title,
  Subtitle,
  Description,
  List,
  ListItem: List.Item,
  Block,
});
