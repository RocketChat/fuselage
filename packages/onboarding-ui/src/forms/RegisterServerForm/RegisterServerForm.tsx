import {
  Box,
  ButtonGroup,
  Button,
  Field,
  EmailInput,
  CheckBox,
  Label,
} from '@rocket.chat/fuselage';
import { useUniqueId, useBreakpoints } from '@rocket.chat/fuselage-hooks';
import { Form, ActionLink } from '@rocket.chat/layout';
import type { ReactElement } from 'react';
import type { SubmitHandler, Validate } from 'react-hook-form';
import { useForm, FormProvider } from 'react-hook-form';
import { useTranslation, Trans } from 'react-i18next';

export type RegisterServerPayload = {
  email: string;
  registerType: 'registered' | 'standalone';
  agreement: boolean;
  updates: boolean;
};

type RegisterServerFormProps = {
  currentStep: number;
  stepCount: number;
  initialValues?: Partial<RegisterServerPayload>;
  validateEmail?: Validate<string>;
  onSubmit: SubmitHandler<RegisterServerPayload>;
  onClickRegisterLater: () => void;
  termsHref?: string;
  policyHref?: string;
  offline?: boolean;
};

const RegisterServerForm = ({
  currentStep,
  stepCount,
  initialValues,
  validateEmail,
  offline,
  onSubmit,
  onClickRegisterLater,
  termsHref = 'https://rocket.chat/terms',
  policyHref = 'https://rocket.chat/privacy',
}: RegisterServerFormProps): ReactElement => {
  const { t } = useTranslation();
  const emailField = useUniqueId();

  const breakpoints = useBreakpoints();
  const isMobile = !breakpoints.includes('md');

  const form = useForm<RegisterServerPayload>({
    mode: 'onChange',
    defaultValues: {
      email: '',
      registerType: 'registered',
      agreement: false,
      updates: true,
      ...initialValues,
    },
  });

  const {
    register,
    formState: { isSubmitting, isValid, errors },
    handleSubmit,
  } = form;

  return (
    <FormProvider {...form}>
      <Form onSubmit={handleSubmit(onSubmit)}>
        <Form.Header>
          <Form.Steps currentStep={currentStep} stepCount={stepCount} />
          <Form.Title>{t('form.registeredServerForm.title')}</Form.Title>
        </Form.Header>
        {!offline && (
          <Form.Container>
            <Field>
              <Field.Label
                required
                display='flex'
                alignItems='center'
                htmlFor={emailField}
              >
                {t('form.registeredServerForm.fields.accountEmail.inputLabel')}
              </Field.Label>
              <Field.Row>
                <EmailInput
                  {...register('email', {
                    required: true,
                    validate: validateEmail,
                  })}
                  placeholder={t(
                    'form.registeredServerForm.fields.accountEmail.inputPlaceholder'
                  )}
                  id={emailField}
                />
              </Field.Row>
              {errors.email && (
                <Field.Error>{t('component.form.requiredField')}</Field.Error>
              )}
            </Field>
            <Box mbs={24}>
              <Box
                mbe={8}
                display='flex'
                flexDirection='row'
                alignItems='flex-start'
                fontScale='c1'
                lineHeight={20}
              >
                <Box mie={8}>
                  <CheckBox {...register('updates')} />{' '}
                </Box>
                <Box is='label' htmlFor='updates'>
                  {t('form.registeredServerForm.keepInformed')}
                </Box>
              </Box>
              <Box
                display='flex'
                flexDirection='row'
                alignItems='flex-start'
                color='default'
                fontScale='c1'
                lineHeight={20}
              >
                <Box mie={8}>
                  <CheckBox {...register('agreement', { required: true })} />{' '}
                </Box>
                <Label htmlFor='agreement' withRichContent required>
                  <Trans i18nKey='component.form.termsAndConditions'>
                    I agree with
                    <a
                      href={termsHref}
                      target='_blank'
                      rel='noopener noreferrer'
                    >
                      Terms and Conditions
                    </a>
                    and
                    <a
                      href={policyHref}
                      target='_blank'
                      rel='noopener noreferrer'
                    >
                      Privacy Policy
                    </a>
                  </Trans>
                </Label>
              </Box>
            </Box>
          </Form.Container>
        )}
        {offline && (
          <Form.Container>
            <Box mbs={32} fontScale='c1' withRichContent>
              {t('form.registeredServerForm.notConnectedToInternet')}
            </Box>
          </Form.Container>
        )}
        <Form.Footer>
          <Box display='flex' flexDirection='column'>
            <ButtonGroup vertical={isMobile} flexGrow={1}>
              <Button type='submit' primary disabled={isSubmitting || !isValid}>
                {offline
                  ? t('component.form.action.registerNow')
                  : t('component.form.action.register')}
              </Button>
              {offline && (
                <ButtonGroup flexGrow={1} align='end' withTruncatedText>
                  <ActionLink onClick={onClickRegisterLater}>
                    {t('form.registeredServerForm.registerLater')}
                  </ActionLink>
                </ButtonGroup>
              )}
            </ButtonGroup>
            <Box mbs={24} fontScale='c1'>
              {t('form.registeredServerForm.registrationEngagement')}
            </Box>
          </Box>
        </Form.Footer>
      </Form>
    </FormProvider>
  );
};

export default RegisterServerForm;
