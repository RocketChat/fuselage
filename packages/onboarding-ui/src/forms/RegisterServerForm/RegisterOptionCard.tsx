import { Divider, CheckBox, RadioButton, Box } from '@rocket.chat/fuselage';
import React, { ReactElement, useEffect } from 'react';
import { useFormContext, useWatch } from 'react-hook-form';
import { useTranslation, Trans } from 'react-i18next';

import OptionCard from './OptionCard';

const RegisterOptionCard = (): ReactElement => {
  const { t } = useTranslation();

  const { register, setValue, trigger } = useFormContext();

  const onClickCard = () => {
    setValue('registerType', 'registered');
    trigger('agreement');
  };

  const selected = useWatch({ name: 'registerType' }) === 'registered';

  useEffect(() => {
    if (!selected) {
      setValue('agreement', false);
      setValue('updates', false);
    }
  }, [selected, setValue]);

  return (
    <OptionCard selected={selected} onClick={onClickCard}>
      <OptionCard.Block>
        <OptionCard.Title>
          <RadioButton
            {...register('registerType')}
            mie='x8'
            value='registered'
          />
          {t('form.serverRegistrationForm.register.title')}
        </OptionCard.Title>
      </OptionCard.Block>
      <OptionCard.Block>
        <OptionCard.Description>
          {t('form.serverRegistrationForm.register.description')}
        </OptionCard.Description>
      </OptionCard.Block>
      <OptionCard.Block>
        <OptionCard.Subtitle>
          {t('form.serverRegistrationForm.register.subtitle')}
        </OptionCard.Subtitle>
        <OptionCard.List>
          <OptionCard.ListItem icon='check'>
            {t('form.serverRegistrationForm.register.included.push')}
          </OptionCard.ListItem>
          <OptionCard.ListItem icon='check'>
            {t('form.serverRegistrationForm.register.included.livechat')}
          </OptionCard.ListItem>
          <OptionCard.ListItem icon='check'>
            {t('form.serverRegistrationForm.register.included.oAuth')}
          </OptionCard.ListItem>
          <OptionCard.ListItem icon='check'>
            {t('form.serverRegistrationForm.register.included.apps')}
          </OptionCard.ListItem>
        </OptionCard.List>
      </OptionCard.Block>
      <Divider />
      <OptionCard.Block>
        <Box mbe='x8' display='block' color='info' fontScale='c1'>
          <CheckBox mie='x8' disabled={!selected} {...register('updates')} />{' '}
          <label htmlFor='updates'>
            {t('form.serverRegistrationForm.register.includeUpdates')}
          </label>
        </Box>
        <Box color='default' fontScale='c1'>
          <CheckBox
            mie='x8'
            disabled={!selected}
            {...register('agreement', {
              validate: (value) => {
                if (!selected) {
                  return true;
                }

                return value;
              },
            })}
          />{' '}
          <Box is='label' htmlFor='agreement' withRichContent>
            <Trans i18nKey='form.serverRegistrationForm.register.agreement'>
              I agree with
              <a
                href='https://rocket.chat/terms'
                target='_blank'
                rel='noopener noreferrer'
              >
                Terms and Conditions
              </a>
              and
              <a
                href='https://rocket.chat/policy'
                target='_blank'
                rel='noopener noreferrer'
              >
                Privacy Policy
              </a>
            </Trans>
          </Box>
        </Box>
      </OptionCard.Block>
    </OptionCard>
  );
};

export default RegisterOptionCard;
