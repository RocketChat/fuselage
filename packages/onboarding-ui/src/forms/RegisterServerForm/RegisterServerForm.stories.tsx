import type { Meta, Story } from '@storybook/react';
import type { ComponentProps } from 'react';

import RegisterServerForm from './RegisterServerForm';

type Args = ComponentProps<typeof RegisterServerForm>;

export default {
  title: 'forms/RegisterServerForm',
  component: RegisterServerForm,
  parameters: {
    layout: 'centered',
    actions: { argTypesRegex: '^on.*' },
  },
  args: {
    currentStep: 1,
    stepCount: 1,
    agreementHref: '#',
    policyHref: '#',
  },
} as Meta<Args>;

export const _RegisterServerForm: Story<Args> = (args) => (
  <RegisterServerForm {...args} />
);
_RegisterServerForm.storyName = 'RegisterServerForm';
