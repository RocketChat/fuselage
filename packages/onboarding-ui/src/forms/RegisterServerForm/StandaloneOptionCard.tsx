import { RadioButton } from '@rocket.chat/fuselage';
import type { ReactElement } from 'react';
import { useFormContext, useWatch } from 'react-hook-form';
import { useTranslation } from 'react-i18next';

import OptionCard from './OptionCard';

const RegisterOptionCard = (): ReactElement => {
  const { t } = useTranslation();

  const { register, setValue, trigger } = useFormContext();

  const onClickCard = () => {
    setValue('registerType', 'standalone');
    trigger('agreement');
  };

  const selected = useWatch({ name: 'registerType' }) === 'standalone';

  return (
    <OptionCard selected={selected} onClick={onClickCard}>
      <OptionCard.Block>
        <OptionCard.Title>
          <RadioButton
            {...register('registerType')}
            mie='x8'
            value='standalone'
          />
          {t('form.serverRegistrationForm.standalone.title')}
        </OptionCard.Title>
      </OptionCard.Block>
      <OptionCard.Block>
        <OptionCard.Description>
          {t('form.serverRegistrationForm.standalone.description')}
        </OptionCard.Description>
        <OptionCard.List listStyle='inside' spacing='x2' color='info'>
          <OptionCard.ListItem>
            {t('form.serverRegistrationForm.standalone.actionList.accounts')}
          </OptionCard.ListItem>
          <OptionCard.ListItem>
            {t('form.serverRegistrationForm.standalone.actionList.settings')}
          </OptionCard.ListItem>
          <OptionCard.ListItem>
            {t('form.serverRegistrationForm.standalone.actionList.mobile')}
          </OptionCard.ListItem>
        </OptionCard.List>
      </OptionCard.Block>
    </OptionCard>
  );
};

export default RegisterOptionCard;
