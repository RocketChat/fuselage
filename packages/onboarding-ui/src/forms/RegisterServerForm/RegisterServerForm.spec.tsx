import ReactDOM from 'react-dom';

import RegisterServerForm from './RegisterServerForm';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <RegisterServerForm
      currentStep={1}
      stepCount={1}
      validateEmail={() => true}
      onClickRegisterLater={() => undefined}
      onSubmit={() => undefined}
    />,
    div
  );
  ReactDOM.unmountComponentAtNode(div);
});
