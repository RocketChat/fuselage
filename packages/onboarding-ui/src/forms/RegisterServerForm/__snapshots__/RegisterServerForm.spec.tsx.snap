// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders _RegisterServerForm without crashing 1`] = `
<body>
  <div>
    <form
      aria-describedby=":r0:-informed-disclaimer :r0:-engagement-disclaimer"
      aria-labelledby=":r0:-title"
      class="rcx-box rcx-box--full rcx-tile--elevation-1 rcx-tile rcx-css-vuqkc1"
      tabindex="-1"
    >
      <header
        class="rcx-box rcx-box--full rcx-css-heem01"
      >
        <div
          class="rcx-box rcx-box--full rcx-css-141nwsj"
        >
          Step 1 of 1
        </div>
        <div
          class="rcx-box rcx-box--full rcx-css-18gg2kh"
          id=":r0:-title"
        >
          Register your workspace
        </div>
      </header>
      <div
        class="rcx-box rcx-box--full"
      >
        <fieldset
          class="rcx-box rcx-box--full rcx-field-group"
          role="group"
        >
          <div
            class="rcx-box rcx-box--full rcx-field rcx-field-group__item"
          >
            <label
              class="rcx-box rcx-box--full rcx-label rcx-box rcx-box--full rcx-field__label rcx-css-127j9mz"
              for=":r1:"
            >
              Admin email
              <span
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-label__required rcx-css-b7g1a9"
              >
                *
              </span>
            </label>
            <span
              class="rcx-box rcx-box--full rcx-field__row"
            >
              <input
                aria-describedby=":r1:-error"
                aria-invalid="false"
                aria-required="true"
                class="rcx-box rcx-box--full rcx-box--animated rcx-input-box--type-email rcx-input-box"
                id=":r1:"
                name="email"
                placeholder="Insert your email to continue"
                size="1"
                type="email"
                value=""
              />
            </span>
          </div>
          <div
            class="rcx-box rcx-box--full rcx-field rcx-css-6qd6i7"
          >
            <span
              class="rcx-box rcx-box--full rcx-field__row rcx-css-gj79e2"
            >
              <label
                class="rcx-box rcx-box--full rcx-check-box"
              >
                <input
                  aria-describedby=":r2:-error"
                  aria-invalid="false"
                  aria-required="true"
                  class="rcx-box rcx-box--full rcx-check-box__input"
                  id=":r2:"
                  name="agreement"
                  type="checkbox"
                />
                <i
                  aria-hidden="true"
                  class="rcx-box rcx-box--full rcx-check-box__fake"
                />
              </label>
              <label
                class="rcx-box rcx-box--full rcx-label rcx-box rcx-box--full rcx-box--with-block-elements rcx-box--with-inline-elements rcx-field__label rcx-css-1nlo49l"
                for=":r2:"
              >
                I agree with 
                <a
                  class="rcx-box rcx-box--full"
                  href="https://rocket.chat/terms"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Terms and Conditions
                </a>
                 and 
                <a
                  class="rcx-box rcx-box--full"
                  href="https://rocket.chat/privacy"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Privacy Policy
                </a>
                <span
                  aria-hidden="true"
                  class="rcx-box rcx-box--full rcx-label__required rcx-css-b7g1a9"
                >
                  *
                </span>
              </label>
            </span>
          </div>
          <input
            class="rcx-field-group__item"
            name="updates"
            type="hidden"
            value="true"
          />
        </fieldset>
      </div>
      <footer
        class="rcx-box rcx-box--full rcx-css-c21cfe"
      >
        <div
          class="rcx-box rcx-box--full rcx-css-1dl0zzg"
        >
          <div
            class="rcx-button-group rcx-button-group--vertical rcx-button-group--align-start"
            role="group"
          >
            <button
              class="rcx-box rcx-box--full rcx-button--primary rcx-button rcx-button-group__item"
              type="submit"
            >
              <span
                class="rcx-button--content"
              >
                Register workspace
              </span>
            </button>
          </div>
          <div
            class="rcx-box rcx-box--full rcx-css-19c5bc1"
            id=":r0:-engagement-disclaimer"
          >
            Registration allows automatic license updates, notifications of critical vulnerabilities and access to Rocket.Chat Cloud services. No sensitive workspace data is shared; statistics sent to Rocket.Chat are made visible to you within the administration area.
          </div>
          <div
            class="rcx-box rcx-box--full rcx-css-19c5bc1"
            id=":r0:-informed-disclaimer"
          >
            By submitting this form you consent to receive more information about Rocket.Chat products, events and updates, according to our 
            <a
              href="https://rocket.chat/privacy"
              rel="noopener noreferrer"
              target="_blank"
            >
              privacy policy
            </a>
            . You may unsubscribe at any time.
          </div>
        </div>
      </footer>
    </form>
  </div>
</body>
`;

exports[`renders _RegisterServerFormOffline without crashing 1`] = `
<body>
  <div>
    <form
      aria-describedby=":r3:-informed-disclaimer :r3:-engagement-disclaimer"
      aria-labelledby=":r3:-title"
      class="rcx-box rcx-box--full rcx-tile--elevation-1 rcx-tile rcx-css-vuqkc1"
      tabindex="-1"
    >
      <header
        class="rcx-box rcx-box--full rcx-css-heem01"
      >
        <div
          class="rcx-box rcx-box--full rcx-css-141nwsj"
        >
          Step 1 of 1
        </div>
        <div
          class="rcx-box rcx-box--full rcx-css-18gg2kh"
          id=":r3:-title"
        >
          Register your workspace
        </div>
      </header>
      <div
        class="rcx-box rcx-box--full"
      >
        <fieldset
          class="rcx-box rcx-box--full rcx-field-group"
          role="group"
        >
          <div
            class="rcx-box rcx-box--full rcx-field rcx-field-group__item"
          >
            <label
              class="rcx-box rcx-box--full rcx-label rcx-box rcx-box--full rcx-field__label rcx-css-127j9mz"
              for=":r4:"
            >
              Admin email
              <span
                aria-hidden="true"
                class="rcx-box rcx-box--full rcx-label__required rcx-css-b7g1a9"
              >
                *
              </span>
            </label>
            <span
              class="rcx-box rcx-box--full rcx-field__row"
            >
              <input
                aria-describedby=":r4:-error"
                aria-invalid="false"
                aria-required="true"
                class="rcx-box rcx-box--full rcx-box--animated rcx-input-box--type-email rcx-input-box"
                id=":r4:"
                name="email"
                placeholder="Insert your email to continue"
                size="1"
                type="email"
                value=""
              />
            </span>
          </div>
          <div
            class="rcx-box rcx-box--full rcx-field rcx-css-6qd6i7"
          >
            <span
              class="rcx-box rcx-box--full rcx-field__row rcx-css-gj79e2"
            >
              <label
                class="rcx-box rcx-box--full rcx-check-box"
              >
                <input
                  aria-describedby=":r5:-error"
                  aria-invalid="false"
                  aria-required="true"
                  class="rcx-box rcx-box--full rcx-check-box__input"
                  id=":r5:"
                  name="agreement"
                  type="checkbox"
                />
                <i
                  aria-hidden="true"
                  class="rcx-box rcx-box--full rcx-check-box__fake"
                />
              </label>
              <label
                class="rcx-box rcx-box--full rcx-label rcx-box rcx-box--full rcx-box--with-block-elements rcx-box--with-inline-elements rcx-field__label rcx-css-1nlo49l"
                for=":r5:"
              >
                I agree with 
                <a
                  class="rcx-box rcx-box--full"
                  href="https://rocket.chat/terms"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Terms and Conditions
                </a>
                 and 
                <a
                  class="rcx-box rcx-box--full"
                  href="https://rocket.chat/privacy"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Privacy Policy
                </a>
                <span
                  aria-hidden="true"
                  class="rcx-box rcx-box--full rcx-label__required rcx-css-b7g1a9"
                >
                  *
                </span>
              </label>
            </span>
          </div>
          <input
            class="rcx-field-group__item"
            name="updates"
            type="hidden"
            value="true"
          />
        </fieldset>
      </div>
      <footer
        class="rcx-box rcx-box--full rcx-css-c21cfe"
      >
        <div
          class="rcx-box rcx-box--full rcx-css-1dl0zzg"
        >
          <div
            class="rcx-button-group rcx-button-group--vertical rcx-button-group--align-start"
            role="group"
          >
            <button
              class="rcx-box rcx-box--full rcx-button--primary rcx-button rcx-button-group__item"
              disabled=""
              type="submit"
            >
              <span
                class="rcx-button--content"
              >
                Register workspace
              </span>
            </button>
            <button
              class="rcx-box rcx-box--full rcx-button rcx-button-group__item"
              type="button"
            >
              <span
                class="rcx-button--content"
              >
                Register offline
              </span>
            </button>
          </div>
          <div
            class="rcx-box rcx-box--full rcx-css-19c5bc1"
            id=":r3:-engagement-disclaimer"
          >
            Registration allows automatic license updates, notifications of critical vulnerabilities and access to Rocket.Chat Cloud services. No sensitive workspace data is shared; statistics sent to Rocket.Chat are made visible to you within the administration area.
          </div>
          <div
            class="rcx-box rcx-box--full rcx-css-19c5bc1"
            id=":r3:-informed-disclaimer"
          >
            By submitting this form you consent to receive more information about Rocket.Chat products, events and updates, according to our 
            <a
              href="https://rocket.chat/privacy"
              rel="noopener noreferrer"
              target="_blank"
            >
              privacy policy
            </a>
            . You may unsubscribe at any time.
          </div>
        </div>
      </footer>
    </form>
  </div>
</body>
`;
